### PR TITLE
Add support for source-build

### PR DESCRIPTION
**Pull request by JunTaoLuo**
Addresses #3752

### CODE REVIEW

File: eng/AfterSolutionBuild.targets [hunk 1]
```diff
@@ -17,7 +17,7 @@
       SharedFrameworkTargetFramework="netcoreapp$(AspNetCoreMajorVersion).$(AspNetCoreMinorVersion)" />
   </Target>
 
-  <Target Name="GenerateBuildAssetManifest" AfterTargets="Pack" Condition="'$(ContinuousIntegrationBuild)' == 'true'">
+  <Target Name="GenerateBuildAssetManifest" AfterTargets="Pack" Condition="'$(ContinuousIntegrationBuild)' == 'true' and '$(DotNetBuildFromSource)' != 'true'">
```
Comment (JunTaoLuo) [inline eng/AfterSolutionBuild.targets]: I think we should skip publishing source build artifacts to BAR.

**Comment by JunTaoLuo** [comment]
@aspnet-hello triage https://dev.azure.com/dnceng/public/_build/results?buildId=253104&view=logs

**Comment by aspnet-hello** [comment]
This comment was made automatically. If there is a problem contact aspnetcore-build@microsoft.com.

I've triaged the above build. I've created/commented on the following issue(s)
https://github.com/aspnet/AspNetCore-Internal/issues/2771

File: eng/AfterSolutionBuild.targets (hunk 1 above)
Comment (dougbu) [reply eng/AfterSolutionBuild.targets]: Agreed

File: .azure/pipelines/ci.yml [hunk 2]
```diff
@@ -471,3 +471,50 @@ jobs:
...
+    condition: always()
+    continueOnError: true
+    inputs:
+      pathtoPublish: artifacts/log/
+      artifactName: Source_Build_Logs
+      artifactType: Container
+      parallel: true
+  - task: PublishBuildArtifacts@1
+    displayName: Upload package artifacts
+    condition: always()
```
Comment (dougbu) [inline .azure/pipelines/ci.yml]: Never upload packages for builds from a fork. See for example the equivalent in Extensions:
```suggestion
    # Only capture source build artifacts in PRs for the sake of inspecting
    # changes that impact source-build. The artifacts from this build pipeline are never actually used.
    condition: and(succeeded(), eq(variables['Build.Reason'], 'PullRequest'))
```

File: eng/Workarounds.targets [hunk 3]
```diff
@@ -21,10 +21,15 @@
   </ItemGroup>
 
   <!-- Workaround https://github.com/aspnet/AspNetCore/issues/7503. This chains GenerateSourceLinkFile before razor component targets run. -->
-  <Target Name="_EnsureSourceLinkHappensBeforeRazorComponentGeneration"
+  <!-- Workaround https://github.com/dotnet/source-build/issues/1112. Source link is currently disabled in source build so do not apply this worksaround. -->
```
Comment (dougbu) [inline eng/Workarounds.targets]: Should either restore the link to our tracking bug (https://github.com/aspnet/AspNetCore/issues/7503) or include it too.

Also, restore comments describing what the target does.

File: eng/WorkaroundsImported.targets [hunk 4]
```diff
@@ -0,0 +1,5 @@
+<!-- Use this file to workaround issues. List the issue tracking the item to fix so we can remove the workaround when the issue is resolved. -->
+<Project>
+  <!-- Workaround https://github.com/dotnet/source-build/issues/1112. Source link is currently disabled in source build so define this dummy target which is required for pack. -->
+  <Target Name="InitializeSourceControlInformation"/>
```
Comment (dougbu) [inline eng/WorkaroundsImported.targets]: This is an empty target in Microsoft.Common.CurrentVersion.targets. Why does it need to be defined again here?

File: eng/scripts/ci-source-build.sh [hunk 5]
```diff
@@ -0,0 +1,39 @@
+#!/usr/bin/env bash
```
Comment (dougbu) [inline eng/scripts/ci-source-build.sh]: Is this file identical to the one in Extensions?

File: src/Framework/ref/Microsoft.AspNetCore.App.Ref.csproj [hunk 6]
```diff
@@ -76,6 +76,9 @@ This package is an internal implementation of the .NET Core SDK and is not meant
       _InstallTargetingPackIntoLocalDotNet;
       _CreateTargetingPackArchive;
     </BuildDependsOn>
+
+    <!-- No-op when in source build -->
+    <BuildDependsOn Condition="'$(IsTargetingPackBuilding)' != 'false' and '$(DotNetBuildFromSource)' == 'true'"/>
```
Comment (dougbu) [inline src/Framework/ref/Microsoft.AspNetCore.App.Ref.csproj]: Delete because this additional definition of the target should override the previous one. Not needed anyhow.

File: src/Identity/UI/src/Microsoft.AspNetCore.Identity.UI.csproj [hunk 7]
```diff
@@ -40,7 +40,8 @@
     <UIFrameworkVersionMoniker Include="V4" />
   </ItemGroup>
 
-  <Target Name="BuildRazorViews" DependsOnTargets="Compile" BeforeTargets="Build">
+  <!-- Source build doesn't build this package -->
+  <Target Name="BuildRazorViews" DependsOnTargets="Compile" BeforeTargets="Build" Condition="'$(DotNetBuildFromSource)' != 'true'">
```
Comment (dougbu) [inline src/Identity/UI/src/Microsoft.AspNetCore.Identity.UI.csproj]: Why is this necessary? This project isn't in Microsoft.AspNetCore.App

**Review by dougbu — CHANGES_REQUESTED** [review]

File: eng/Workarounds.targets (hunk 3 above)
Comment (JunTaoLuo) [reply eng/Workarounds.targets]: I didn't remove anything?

File: eng/WorkaroundsImported.targets (hunk 4 above)
Comment (JunTaoLuo) [reply eng/WorkaroundsImported.targets]: It's normally a target defined by in source link. I see that in a few place in our code we have custom targets that depends on this target or reference it in Before/AfterTarget during pack. Since sourcelink is currently disabled, we need to define a this target so that those custom targets still work.

File: eng/scripts/ci-source-build.sh (hunk 5 above)
Comment (JunTaoLuo) [reply eng/scripts/ci-source-build.sh]: Not quite, I had to add a step to build the RepoTask.csproj.

File: src/Framework/ref/Microsoft.AspNetCore.App.Ref.csproj (hunk 6 above)
Comment (JunTaoLuo) [reply src/Framework/ref/Microsoft.AspNetCore.App.Ref.csproj]: I don't quite follow? Removing this line breaks things. By default, this is not empty and there are targets that run.

File: src/Identity/UI/src/Microsoft.AspNetCore.Identity.UI.csproj (hunk 7 above)
Comment (JunTaoLuo) [reply src/Identity/UI/src/Microsoft.AspNetCore.Identity.UI.csproj]: So what happens is that when `ExcludeFromSourceBuild` is set is that noop'ing targets are imported and run during build: https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.Arcade.Sdk/tools/Empty.targets. However, the way that these targets are defined, they still run explicitly when these projects are built, even though the Build target itself noops.

File: .azure/pipelines/ci.yml (hunk 2 above)
Comment (JunTaoLuo) [reply .azure/pipelines/ci.yml]: Actually, how about I set it to just Build.Reason = PullRequest. I was debugging earlier and the succeed() condition meant I couldn't get the artifacts.

Comment (dougbu) [reply .azure/pipelines/ci.yml]: That's fine while you're debugging but not once this is checked in. Look at all other artifacts published in this pipeline. Only artifacts/log files are ever published on error.

File: eng/Workarounds.targets (hunk 3 above)
Comment (dougbu) [reply eng/Workarounds.targets]: Not sure why but I thought line 23 was removed.

:ok: just make it clear #7503 tracks dotnet/source-build#1112

File: src/Framework/ref/Microsoft.AspNetCore.App.Ref.csproj (hunk 6 above)
Comment (dougbu) [reply src/Framework/ref/Microsoft.AspNetCore.App.Ref.csproj]: Is that coming in another iteration?

**Review by dougbu — APPROVED** [review]
Just a couple of outstanding points, including a few on earlier iterations

File: .azure/pipelines/ci.yml (hunk 2 above)
Comment (dougbu) [reply .azure/pipelines/ci.yml]: Oh, you've left `succeeded()` in. Alright&hellip;

Comment (JunTaoLuo) [reply .azure/pipelines/ci.yml]: I made the comment after merging your suggestion, wanted to see what you think before I make the changes.

File: eng/Workarounds.targets (hunk 3 above)
Comment (JunTaoLuo) [reply eng/Workarounds.targets]: I don't think it's quite the same. The issue described in #7503 is general to source link. dotnet/source-build#1112 is specific to source build.

File: src/Framework/ref/Microsoft.AspNetCore.App.Ref.csproj (hunk 6 above)
Comment (JunTaoLuo) [reply src/Framework/ref/Microsoft.AspNetCore.App.Ref.csproj]: I removed the condition above but will keep this one.

**Comment by dougbu** [comment]
Hurray:grey_exclamation:

**Comment by sharpninja** [comment]
What is the expected happy path for building the repository from scratch on Ubuntu 18.04 in WSL 2?  I'm using this script:

```bash
#!/usr/bin/env bash

cd /mnt/e/GitHub/sharpninja

rm -r -f aspnetcore

git clone --recursive git@github.com:sharpninja/AspNetCore.git

git submodule update --init --recursive

cd aspnetcore

git remote add upstream https://github.com/aspnet/aspnetcore
git fetch upstream
git merge upstream/master

./restore.sh
./build.sh --all --restore -c Debug -v d
```
Unfortunately, it attempts to import `internal.aspnetcore.buildtasks/3.0.0-build-20190530.3` twice and bombs out on the second attempt.  I'm attaching a log file.
[aspnet-build.log](https://github.com/aspnet/AspNetCore/files/3477459/aspnet-build.log)

**Comment by dougbu** [comment]
@sharpninja it sounds like you're having an issue with building this repo locally. If that's the case, commenting on a merged PR is not likely to get attention. Please file a new issue if my comments don't help.

Suggest leaving `--restore` out of the `./build.sh` command. In addition, execute `git submodule update --init --recursive` much later, after updating your local 'master' branch and just before `./restore.sh`.

In case your fork contains commits that are causing problems, I also suggest `git reset --hard upstream/master` instead of `git merge upstream/master`.
